### ISSUE TITLE
Unable to manipulate compound type properties in WireData object

### ISSUE THREAD

**Issue by KimPeeters** [ryancramerdesign/ProcessWire#1298]
I noticed i couldn't alter an array/object stored in an object based on WireData.

Example:

```php
<?php
require 'index.php';

$data = new WireData();
$data->customArray = array('key1' => 'value1');
$data->customeArray['key2']  = 'key2';

var_dump($data->customArray);
// returns:
// array (size=1)
//   'key1' => string 'value1' (length=6)
```

The functions WireData::__get and WireData::get return a copied version of the value, so therefor altering them doesn't work. Returning them by reference does work.

This doesn't effect scalar values because theses are overridden upon alteration (by WireData::__set).

Any thoughts on this?

Regards, Kim

**Comment by LostKobrakai** [comment]
If I remember correctly this is expected behavior, but maybe Ryan has some more explanatory answer to that. Anyway you should remove the typo `$data->customArray` vs. `$data->customeArray`

**Comment by KimPeeters** [comment]
Thx. Fixed it.

**Comment by ryancramerdesign** [comment]
That's correct, we don't allow direct manipulation of array values. Maybe we should, and something to think about more, but I thought it safer not to. Besides, PHP always gives a notice (in debug mode) when you try to do that.

**Comment by KimPeeters** [comment]
Indeed it gives a notices. It might be safer but is also unexpected. When defining an array, whether this is in an object or not, you'd expect it to behave as an array.

That said, i have no idea what impact this would have on the working of ProcessWire because it heavily depends on the magic get function in the Wire classes. To test i've replaced all function declarations for __get($key) and get($key) with &__get($key) and &get($key) and it keeps working like a charm but not sure if this is the way to go.

**Comment by ryancramerdesign** [comment]
Looking closer at it, I think it's probably not a good thing for us to do here, except perhaps in rare/specific cases. It violates object encapsulation, providing direct access to properties as if they were public. As a tangible example of why that might be a problem, consider the corresponding set() and __set() methods. They have no opportunity to validate values that are set to the object or modified in the object. There's no way for an object to detect when a change has occurred either. I think there's potential for major problems if objects expose direct references to their properties like this. 

> It might be safer but is also unexpected. When defining an array, whether this is in an object or not, you'd expect it to behave as an array.

I don't necessarily agree that it's unexpected. It behaves exactly how I would expect it to at least. :) ProcessWire doesn't have any public properties on any of its classes, and those are the only things that would behave in that "direct reference" manner. PW sticks to PHP's defaults when it comes to what is, and isn't, passed by reference – the only thing PHP passes by reference are objects. 

However, I do agree that there are some times where it would be convenient to be able to directly manipulate an array, for example. But we'd really be breaking OO rules to get there, and have a far more error prone architecture. The right way to accommodate that particular need is to pass around objects that behave as arrays (like WireArrays, that can validate and change track themselves). But it's also less efficient than just using regular arrays, so a tradeoff. I think where we are is probably a good place to stay…

But there may be a few rare instances where it would be okay for us to provide a reference to an array. For instance, the commonly used HookEvent::arguments array ($event->arguments) may be okay, since it's not necessary for the system to keep track of changes to it, and the only reason it's provided to hooks is specifically so that they can inspect or change it if they want to. There may be other similar examples, though can't think of any at the moment, but will keep it in mind as a potential optimization in places like this.

**Comment by KimPeeters** [comment]
Thanks for the detailed response and I see your point. Adjusting this I would result in undesired situations and behavior. I doubt indeed how often this is a problem and for me the only occasion where i ran in to this was in the config.php.

```php
$config->httpHosts = array('www.site.nl');
if ( $site_env == 'dev' ) {
  $config->httpHosts[] = $_SERVER['HTTP_HOST'];
}
```
